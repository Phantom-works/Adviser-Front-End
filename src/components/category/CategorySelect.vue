<template>
  <div class="categories-container">
    <category-input-field @AddCategory="AddCategory"/>
    <CategoryList :categories="categories" @DeleteCategory="DeleteCategory" />
  </div>
</template>

<script lang="ts" setup>
import CategoryList from '@/components/category/CategoryList.vue'
import type Category from '@/Model/Category'

import { ref } from 'vue'
import CategoryInputField from '@/components/category/CategoryInputField.vue'
function DeleteCategory(_id: number) {
  categories.value = categories.value.filter((category) => category.id !== _id)
}

function AddCategory(_title: string) {
  let category = ref<Category>({
    id: categories.value.length + 1,
    title: _title
  })

  categories.value.push(category.value)
}

let categories = ref<Category[]>([
])
</script>

<style scoped>
.categories-container{
  width: 100%;
}
</style>

