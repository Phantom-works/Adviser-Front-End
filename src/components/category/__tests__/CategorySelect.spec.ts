import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import type Category from '../../../Model/Category'
import CategorySelect from '../CategorySelect.vue'
import { ref } from 'vue'
describe('CategorySelect Snapshot tests', () => {
  // Arrange

  it('Snapshot Empty List', () => {
    //Arrange
    const categories = ref<Category[]>([])

    // Act
    const wrapper = mount(CategorySelect, {
      props: {
        categories: categories
      }
    })
    // Assert
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('CategorySelect Component tests', () => {
  // Arrange

  it('AddCategory No Category inputted, amount 0', async () => {
    // Arrange
    // Negate jest from giving error when window.alert is called

    window.alert = () => {}

    const categories = ref<Category[]>([])

    const wrapper = mount(CategorySelect, {
      props: {
        categories: categories
      }
    })

    // Act

    const button = wrapper.find('button')
    await button.trigger('click.left')

    // Assert
    expect(wrapper.vm.categories).toHaveLength(0)
  })

  it('AddCategory Category, input: title = Test, amount = 1', async () => {
    // Arrange
    const categories = ref<Category[]>([])

    const wrapper = mount(CategorySelect, {
      props: {
        categories: categories
      }
    })

    // Act

    const button = wrapper.find('button')
    const input = wrapper.find('input')
    await input.setValue('Test')
    await button.trigger('click.left')

    // Assert
    expect(wrapper.vm.categories).toHaveLength(1)
    expect(wrapper.vm.categories[0].id).toBe(1)
    expect(wrapper.vm.categories[0].title).toBe('Test')
  })

  it('Add and remove Category, Value: id = 1, title = Test', async () => {
    // Arrange

    const categories = ref<Category[]>([])

    const wrapper = mount(CategorySelect, {
      props: {
        categories: categories
      }
    })

    // Act
    // Add
    const button = wrapper.find('button')
    const input = wrapper.find('input')
    await input.setValue('Test')
    await button.trigger('click.left')

    // Remove
    const xMark = wrapper.find('i')
    await xMark.trigger('click.left')

    // Assert
    expect(wrapper.vm.categories).toHaveLength(0)
  })
})
