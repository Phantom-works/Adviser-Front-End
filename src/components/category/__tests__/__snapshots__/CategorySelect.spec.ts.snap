// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryList Snapshots > Snapshot 9 categories differentiating text length 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;

exports[`CategoryList Snapshots > Snapshot text empty 3 categories 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object],[object Object],[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;

exports[`CategoryList Snapshots > Snapshot text long 5 categoeries 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object],[object Object],[object Object],[object Object],[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;

exports[`CategorySelect Snapshot tests > Snapshot Empty List 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;

exports[`CategorySelect Snapshots > Snapshot 9 categories differentiating text length 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;

exports[`CategorySelect Snapshots > Snapshot empty list 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;

exports[`CategorySelect Snapshots > Snapshot text empty 3 categories 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object],[object Object],[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;

exports[`CategorySelect Snapshots > Snapshot text long 5 categoeries 1`] = `
"<div data-v-8d32426d=\\"\\" class=\\"categories-container\\" categories=\\"[object Object],[object Object],[object Object],[object Object],[object Object]\\">
  <div data-v-15dafa4e=\\"\\" data-v-8d32426d=\\"\\" class=\\"input-container\\"><input data-v-15dafa4e=\\"\\" class=\\"rounded-input\\" placeholder=\\"Enter text here\\" type=\\"text\\"><button data-v-dec1f997=\\"\\" data-v-15dafa4e=\\"\\" class=\\"button\\">+</button></div>
</div>"
`;
